File: frontend/src/types/trace.ts
/**
 * Types for traceability APIs.
 *
 * ``TraceEvent`` corresponds to a single event in a trace series. ``TraceSummary``
 * summarises a trace for listing endpoints.
 */

export interface TraceEvent {
  id: number;
  trace_id: string;
  ts: string;
  actor: string;
  event_type: string;
  payload: Record<string, any>;
  sha256?: string | null;
  prev_sha256?: string | null;
}

export interface TraceSummary {
  trace_id: string;
  first_ts: string;
  last_ts: string;
  count: number;
}

File: backend/schemas/trace.py
"""Pydantic models for traceability APIs.

These schemas represent both individual trace events and aggregated
summaries of traces. A ``TraceEvent`` mirrors the ORM model while
``TraceSummary`` provides a condensed overview of a trace series for
listing endpoints.
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional, Dict, Any
from pydantic import BaseModel


class TraceEvent(BaseModel):
    """Serialized representation of a single trace event."""

    id: int
    trace_id: str
    ts: datetime
    actor: str
    event_type: str
    payload: Dict[str, Any]
    sha256: Optional[str] = None
    prev_sha256: Optional[str] = None

    model_config = {"from_attributes": True}


class TraceSummary(BaseModel):
    """Aggregated overview of a trace series."""

    trace_id: str
    first_ts: datetime
    last_ts: datetime
    count: int
